akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/RoutingLogic.scala

```diff
@@ -76,14 +82,57 @@ private[akka] object RoutingLogics {
 
     private var currentRoutees: Array[ActorRef[T]] = _
 
-    override def selectRoutee(): ActorRef[T] = {
+    override def selectRoutee(msg: T): ActorRef[T] = {
       val selectedIdx = ThreadLocalRandom.current().nextInt(currentRoutees.length)
       currentRoutees(selectedIdx)
     }
+
     override def routeesUpdated(newRoutees: Set[ActorRef[T]]): Unit = {
       currentRoutees = newRoutees.toArray
     }
+  }
 
+  final class ConsistentHashingLogic[T](
+      virtualNodesFactor: Int,
+      mapping: ConsistentHashMapping[T],
+      system: ActorSystem[T])
+      extends RoutingLogic[T] {
+    require(system ne null, "system argument of ConsistentHashingLogic cannot be null.")
+
+    private var currentRoutees: Set[ActorRef[T]] = Set.empty
+
+    private var consistentHash: ConsistentHash[ActorRef[T]] = ConsistentHash(currentRoutees, virtualNodesFactor)
+
+    override def selectRoutee(msg: T): ActorRef[T] =
+      try {
+        mapping(msg) match {
+          case bytes: Array[Byte] => consistentHash.nodeFor(bytes)
+          case str: String        => consistentHash.nodeFor(str)
+          case x: AnyRef =>
+            val bytes = SerializationExtension(system.toClassic).serialize(x).get
+            consistentHash.nodeFor(bytes)
+        }
+      } catch {
+        case NonFatal(e) =>
+          system.log.warn("Couldn't route message [{}] due to [{}]", msg, e.getMessage)
+          system.deadLetters
+      }
+
+    override def routeesUpdated(newRoutees: Set[ActorRef[T]]): Unit = {
+      val withoutOld = currentRoutees.diff(newRoutees).foldLeft(consistentHash)(_ :- _)
+      consistentHash = newRoutees.diff(currentRoutees).foldLeft(withoutOld)(_ :+ _)
+      currentRoutees = newRoutees
+    }
+  }
+
+  object ConsistentHashingLogic {
+    type ConsistentHashMapping[T] = PartialFunction[T, Any]
```

Comment:
Make the public api a total function instead of a partial and skip the type alias. Since this is internal it can anyways not be part of the public API but have to live elsewhere.

Either we select one of the signatures and only support that (`T => Array[Byte]`) or introduce SAM traits for it (which has an added benefit of making the exact same API possible for Java and Scala)

Reply:
After getting a little bit more into the code, the hashing will always happen within existing `ConsistentHash` object.

I've introduced `TypedSerializer[T]` trait, but I do not have good feeling about that. I really could use some guidance with that! For example, should we use existing akka mechanism for serialization?